### DISCUSSION TITLE
Is it possible that trice can replay uart hex file dumped somewhere else?

### DISCUSSION THREAD

**Discussion by ztliu722** [rokath/trice#408] [Q&A] [answered]
Hi, lokath, great thanks for this lovely wonderful, and yet powerful tool!
Currently trice is running on my windows machine nicely.
However, I am wondering if it works in the following scenario:
1. target board uart is received somewhere else, like a rpi or some other embedded system without trice there. 
2. the uart output raw bytes are recorded into a file.
3. Use this file, trice in my local windows machine, with the help of til.json and li.json, can then replay "trice"
Does trice currently support this scenario? 
I notice that trice can output bin file and replay it. 

Anyway, thnaks for this great trace tool.
Zt.

**Comment by rokath** [comment, marked as answer]
`trice h -l` gives you
```bash
syntax: 'trice sub-command' [params]
sub-command 'l|log': For displaying trice logs coming from port. With "trice log" the trice tool display mode is activated.
example: 'trice l -p COM15 -baud 38400': Display trice log messages from serial port COM15
example: 'trice l': Display flexL data format trice log messages from default source J-LINK over Segger RTT protocol.
example: 'trice l -port ST-LINK -v -s': Shows verbose version information and also the received raw bytes.
  -args string
        Use to pass port specific parameters. The "default" value depends on the used port:
        port "BUFFER": default="0 0 0 0", Option for args is any space separated decimal number byte sequence.
        port "DUMP": default="", Option for args is any space or comma separated byte sequence in hex like "7B 1A ee,88, 5a".
        port "COMn": default="", Unused option for a different driver. (For baud rate settings see -baud.)
        port "FILE": default="trices.raw", Option for args is any file name. Trice retries on EOF.
        port "FILEBUFFER": default="trices.raw", Option for args is any file name. Trice stops on EOF.
```
So this should work:
```bash
trice l -p FILEBUFFER -args="path/to/myTriceLogFile.bin" -i="path/to/til.json" -li="path/to/li.sjon" -hs off
```

To store raw date with the trice tool check for log option `-binaryLogfile`.

The trice tool can also transmit and receive over TCP/IP: [display-server-option](https://github.com/rokath/trice#display-server-option)

**Reply by ztliu722** [reply]
Wow! Didn't expect you to reply so soon!
Currenly I can copy n past a uart dump string and use
`trice l -p DUMP -args="...a long long hex dump string..."`
This already satisfies my requirement. 
Thanks again for this great tool!
Zt.

**Comment by rokath** [comment]
May I ask you why you use hex dump and not binary data? Would it be helpful to add an option like `trice l -p HEXDUMP -args file.txt`? Actually I do not see the need for that.

**Reply by ztliu722** [reply]
In fact, you are right! 
I just found that general tools like teraterm support log binary files, and it works just fine with trice port `FILE` or `FILEBUFFER`.

**Reply by rokath** [reply]
When you can run teraterm on a machine, the trice tool should be usable at this same machine directly. Or is there a special reason?

**Reply by ztliu722** [reply]
A special experment will be conducted on the target board, with some space and cabling constraints. 
The logging system may just be another small embedded system with wireless capability.
Also the experiment will be handled by our partner, the information may not all be directly sharable. 
For now, I just need to make sure the logged uart rawbytes can be replayed later with trice in our machine for further analysis.
Indeed, we may need to enable XTEA functionality next time so that the logging embedded system can directly run trice. 
Zt.

**Reply by rokath** [reply]
XTEA right now went out a bit of focus and may not work right away. In v0.57.0 the target code was changed heavily and is better now. XTEA needs multiple of 8 byte packages. Hopefully I get time to fix that. Please be patient - trice is a spare time project.

**Reply by rokath** [reply]
In v0.60.0 XTEA is active again. There are now 2 tests in different modes: 

- double buffer packing several trice messages together in one package 
  - more RAM usage
  - less transfer bytes
  - more data loss when data disruption 
- ring buffer packing each trice message in a separate package
  - less RAM usage
  - more transfer bytes
  - less data loss when data disruption

**Comment by rokath** [comment]
With v0.58.0 you can get XTEA encrypted data over Segger RTT in direct mode. It is just experimental and will be soon available in deferred mode over UART as well.

**Comment by ztliu722** [comment]
No worry, take your pace.
Trice already helps A LOT!
